test/environment_unittest.cc

```diff
@@ -42,6 +49,59 @@ class TemporaryFile {
  private:
   boost::filesystem::path path_;
 };
+
+// Starts a server in a separate thread, allowing it to choose an
+// available port.
+class FakeServerThread {
+ public:
+  FakeServerThread()
+    : server_(Server::options(handler_).address("127.0.0.1").port("")) {
+    server_.listen();
+    thread_ = std::thread([this] { server_.run(); });
+  }
+
+  void SetResponse(const std::string& path, const std::string& response) {
+    handler_.path_responses[path] = response;
+  }
+
+  std::string HostPort() {
+    return server_.address() + ":" + server_.port();
+  }
+
+  ~FakeServerThread() {
+    server_.stop();
+    thread_.join();
+  }
+
+ private:
+  struct Handler;
+  typedef boost::network::http::server<Handler> Server;
+
+  // Handler that maps paths to response strings.
+  struct Handler {
+    void operator() (Server::request const &request,
+                     Server::connection_ptr connection) {
+      auto it = path_responses.find(request.destination);
+      if (it != path_responses.end()) {
+        connection->set_status(Server::connection::ok);
+        connection->set_headers(std::map<std::string, std::string>({
+            {"Content-Type", "text/plain"},
+        }));
+        connection->write(it->second);
+      } else {
+        // Note: We have to set headers; otherwise, an exception is thrown.
+        connection->set_status(Server::connection::not_found);
+        connection->set_headers(std::map<std::string, std::string>());
+      }
+    }
+    std::map<std::string, std::string> path_responses;
```

Comment:
Should this map live in the `FakeServer` class instead?

Reply:
I think I'd have to pass in a FakeServer reference to the Handler, which seems more complicated.